Add support for running promotions on inheritance projects.

J-cztery opened the pull request:
This pull request contains an implementation of extension point to jenkins-inheritance-project-plugin and some changes to JobPropertyImpl construction to allow for interaction with jenkins-inheritance-project-plugin. 
A new constructor has been added that allows reading promotion processes information from the location specified by an existing JobPropertyImpl but makes the JobPropertyImpl instance (and it's promotion processes) independent otherwise. 

Initially I wanted to keep this extension separate (as a plugin) but this proved too difficult because of the hermetization of JobPropertyImpl and its insistence on loading promotion processes on construction. 

Note that there is a problem in jenkins-inheritance-project-plugin and there is an accompanying fix to jenkins-inheritance-project-plugin plugin. That problem prevents from seeing promotions that are currently running. There might be some problems with getting it merged to mainline, so you might want to use my fork directly https://github.com/J-cztery/jenkins-inheritance-plugin. 

This change introduces optional dependency to jenkins-inheritance-project. 

It supersedes the following pull requests: 
https://github.com/jenkinsci/promoted-builds-plugin/pull/74
https://github.com/jenkinsci/promoted-builds-plugin/pull/51 (thanks David Tanner for inspiration).

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java:
```diff
@@ -120,7 +132,10 @@ public boolean accept(File child) {
-        if(subdirs!=null) {
+        loadProcesses(subdirs);
+    }
+	private void loadProcesses(File[] subdirs) throws IOException {
```
tab/spaces mix

J-cztery replied:
Ok. Thanks.

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java:
```diff
@@ -201,7 +216,11 @@ private void buildActiveProcess() throws IOException {
+            String activeProcessName = getActiveProcessName(processName);
+            if (!activeProcessName.equals(processName)){
+            	p.renameTo(activeProcessName);
+            }
```
This change seems to be unrelated to the issue, but seems to be reasonable. Anyway, the plugin would benefit from a configurable case sensitivity handling like it happens with User IDs.

J-cztery replied:
It is actually related to the issue. Unfortunately p.renameTo results in an attempt to lock the promotion process and ultimately a deadlock.

oleg-nenashev replied:
Agreed

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/inheritance/JobPropertyImplSelector.java:
```diff
@@ -0,0 +1,66 @@
+@Extension(optional=true)
+public class JobPropertyImplSelector extends InheritanceSelector<JobProperty<?>> {
+	private static final long serialVersionUID = 6297336734737164557L;
+	private static Logger logger = Logger.getLogger(JobPropertyImplSelector.class);
```
NIT: could be also final

J-cztery replied:
public final class? Ok.

oleg-nenashev replied:
no, private final static.

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/inheritance/JobPropertyImplSelector.java:
```diff
@@ -0,0 +1,66 @@
+
+/**
+ *  
+ * @author Jacek Tomaka
```
A good practice is to add <code>@since TODO</code> to the Javadoc. Then it will be possible to resolve these TODOs before the relelase

J-cztery replied:
Ok. Will add this. Thanks.

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/inheritance/JobPropertyImplSelector.java:
```diff
@@ -0,0 +1,66 @@
+ */
+@Extension(optional=true)
+public class JobPropertyImplSelector extends InheritanceSelector<JobProperty<?>> {
+	private static final long serialVersionUID = 6297336734737164557L;
```
Was it generated automatically? For new classes <code>1L</code> is enough

oleg-nenashev replied:
Please use spaces instead of tabs

J-cztery replied:
Ok. Will change it.

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/inheritance/JobPropertyImplSelector.java:
```diff
@@ -0,0 +1,66 @@
+	
+	@Override
+	public JobPropertyImpl merge(JobProperty<?> prior, JobProperty<?> latter, InheritanceProject caller){
+		return null;
```
From what I see in in https://github.com/i-m-c/jenkins-inheritance-plugin/blob/master/src/main/java/hudson/plugins/project_inheritance/projects/inheritance/InheritanceSelector.java#L258-L268 , it may cause an NPE in your handleSingleton implementation

J-cztery replied:
obj == null? Will add a guard. Thanks.

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/inheritance/JobPropertyImplSelector.java:
```diff
@@ -0,0 +1,66 @@
+		if (JobPropertyImpl.class.getName().equals(obj.getClass().getName())){
+			return "JobPropertyImpl";
+		}
+		return null;
```
null seems to be safe

J-cztery replied:
Sorry, i do not understand this comment. Please elaborate.

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/inheritance/JobPropertyImplSelector.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	public String getObjectIdentifier(JobProperty<?> obj){
+		if (JobPropertyImpl.class.getName().equals(obj.getClass().getName())){
+			return "JobPropertyImpl";
```
I would recommend to use a unique identifier, because this one may cause a collision with other selectors (common class name). Maybe a full class name is better

J-cztery replied:
Ok.

oleg-nenashev commented:
@J-cztery Thanks for taking an action! the selected approach looks to be valid to me. From what I see the issue in inheritance plugin is not a blocker for this PR. Is it correct?

It would be also great to have a functional test verifying the new functionality

oleg-nenashev commented on src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java:
```diff
@@ -112,7 +116,15 @@ private JobPropertyImpl(StaplerRequest req, JSONObject json) throws Descriptor.F
+            });
+    	
+    	loadProcesses(subdirs);
+    }
```
Mostly duplicates <code>init()</code>. Would be great to merge methods somehow

J-cztery commented:
@oleg-nenashev: Thank you for the code review. I will work on your comments next.

J-cztery commented:
@oleg-nenashev: regarding the test. I struggle to write an unit tests that uses multiple plugins. Any ideas on how to set it up?

J-cztery commented:
Yeah, I have just tested this fix with 1.5.3 version of inheritance build and the only problem i could find is that when the promotion build is running it is not shown on Promotion Status in either build context or project context. But it is shown in the queue. 
The problem is due to the fact that inheritance plugin does not cache properties and generate properties on the fly and it so it generates many JobPropertyImpls....
I do not think this one is critical for the PR.

oleg-nenashev commented:
@J-cztery 
Thanks for the changes. The code looks good to me.
Inheritance plugin was in my personal blacklist for a long time due to functional issues, so unfortunately I'm not very familiar with its testing capabilities. BTW, in <code>@JenkinsRule</code> you will automatically get optional plugins installed by default. So you would be able to write a test there. If you describe the reproduction scenario, I probably can write a sample test.

Would you like to keep the test as a part of this PR? If no, I could merge this commit in order to test the behavior together with other changes.

J-cztery commented:
@oleg-nenashev : 
Thank you for reviewing this PR!
I have currently issues with running tests on Windows as this is my environment of choice. 
[CallStack.txt](https://github.com/jenkinsci/promoted-builds-plugin/files/40974/CallStack.txt)
I will try to get them working but this might take time. 

If you would like to create a test then a scenario is : 
1. Create an inheritance project BASE and add jobPropertyImpl with a promotion. 
2. Create an inheritance project DERIVED and add a project reference (Ordered or Simple) to project BASE. 
3. Now if you build DERIVED and if promotion works (does not crash) then it works ;)
So if you find time to write this test then ok, otherwise i will do it. 

I do observe some functional issues with inheritance plugin as well but i am trying to kill them one by one. The problem i see with this plugin is that maintainer is not very responsive and the project is pretty much dead;(

What are your main concerns regarding the functionality of inheritance plugin?
I have added: 
1. ability to specify node(label) of transient projects(in mating), 
2. View full build flow actually shows full build flow.
3. Block when upstream and downstream project is building is now working. 
4. Got parametrized trigger plugin working so that upstream/downstream relationship can be parameterized.
5. Added support to choose priorities of parents of transient projects(in mating)
6. Changed the way inheritance project decides if it should do full inheritance to blacklisting.
7. Recently added caching of properties (This is the thing that promotion-plugin needs).

oleg-nenashev commented:
@J-cztery 
this is a known issue with Windows tests on old Jenkins versions (locking of tmp dirs). You could run tests locally against new Jenkins core versions, then PR builder will run them against the actual version if you don't commit the version change.

We could also fork the Inheritance project plugin to jenkinsci org if @HedAurabesh wants to handover it. Maybe makes sense to start a thread in jenkinsci-dev and Cc owners there. The plugin is definitely useful, so it it great that somebody works on it.

J-cztery commented:
@oleg-nenashev : Thanks for the hint. I will try to run tests against newer version of Jenkins. I have adapted one of the tests to use inheritance project but it would make sense(?) to run most of them if not all against FreeStyle project type as well as InheritanceProject. This would require making the unit tests slightly more complex...Do you think it would be worth it?

Let's deal with potential fork of inheritance project once this PR is finished.

oleg-nenashev commented:
@J-cztery 
Makes sense to create new tests instead of modifying the existing ones. But feel free to copy as much tests as you want if it allows to increase the coverage :)

For such integrations it also makes sense to create a spot-check in https://github.com/jenkinsci/acceptance-test-harness in order to catch regressions in Jenkins core, but it is optional && can be done after this PR.

oleg-nenashev commented:
Performed manual spot-check and it seems to work. 
Excluding it from the today's release to have a deeper look

J-cztery commented:
Yeah.. I am still working on the unit tests and realized that @DavidTanner's change will still be needed (slightly modified to keep the binary compatibility as you requested) and i am not sure if i will be able to get unit tests working with inheritance-plugin 1.53.
So not releasing it just yet makes sense. I will let you know once i am done with tests.

J-cztery commented:
Ready @oleg-nenashev.

oleg-nenashev commented:
:+1:

oleg-nenashev commented:
@J-cztery thanks a lot for your efforts!
